dot/parachain/spam_slots.go

```diff
@@ -0,0 +1,200 @@
+
+// spamSlots is an implementation of SpamSlots
+type spamSlots struct {
+	slots     map[slotKey]uint32
+	slotsLock sync.RWMutex
+
+	unconfirmed     map[unconfirmedKey]*treeset.Set
+	unconfirmedLock sync.RWMutex
+
+	maxSpamVotes uint32
+}
+
+// byValidatorIndex is a comparator for ValidatorIndex
+func byValidatorIndex(a, b interface{}) int {
+	return int(a.(parachain.ValidatorIndex) - b.(parachain.ValidatorIndex))
+}
+
+func slotsKey(session parachain.SessionIndex, validator parachain.ValidatorIndex) slotKey {
+	return slotKey{
+		session:   session,
+		validator: validator,
+	}
+}
+
+func unconfirmedDisputesKey(session parachain.SessionIndex, candidate common.Hash) unconfirmedKey {
+	return unconfirmedKey{
+		session:   session,
+		candidate: candidate,
+	}
+}
+
+func (s *spamSlots) getSpamCount(session parachain.SessionIndex, validator parachain.ValidatorIndex) (uint32, bool) {
+	s.slotsLock.RLock()
+	defer s.slotsLock.RUnlock()
+	spamCount, ok := s.slots[slotsKey(session, validator)]
+	return spamCount, ok
+}
+
+func (s *spamSlots) getValidators(session parachain.SessionIndex, candidate common.Hash) (*treeset.Set, bool) {
+	s.unconfirmedLock.RLock()
+	defer s.unconfirmedLock.RUnlock()
+	validators, ok := s.unconfirmed[unconfirmedDisputesKey(session, candidate)]
+	return validators, ok
+}
+
+func (s *spamSlots) clearValidators(session parachain.SessionIndex, candidate common.Hash) {
+	s.unconfirmedLock.Lock()
+	defer s.unconfirmedLock.Unlock()
+	delete(s.unconfirmed, unconfirmedDisputesKey(session, candidate))
+}
+
+func (s *spamSlots) clearSlots(session parachain.SessionIndex, validator parachain.ValidatorIndex) {
+	s.slotsLock.Lock()
+	defer s.slotsLock.Unlock()
+	delete(s.slots, slotsKey(session, validator))
+}
+
+func (s *spamSlots) incrementSpamCount(session parachain.SessionIndex, validator parachain.ValidatorIndex) {
+	s.slotsLock.Lock()
+	defer s.slotsLock.Unlock()
+	s.slots[slotsKey(session, validator)]++
+}
+
+func (s *spamSlots) AddUnconfirmed(session parachain.SessionIndex, candidate common.Hash, validator parachain.ValidatorIndex) bool {
+	if spamCount, _ := s.getSpamCount(session, validator); spamCount >= s.maxSpamVotes {
+		return false
+	}
+
+	s.unconfirmedLock.Lock()
+	defer s.unconfirmedLock.Unlock()
+
+	validators, ok := s.unconfirmed[unconfirmedDisputesKey(session, candidate)]
+	if !ok || validators == nil {
+		validators = treeset.NewWith(byValidatorIndex)
+	}
+
+	if !validators.Contains(validator) {
+		validators.Add(validator)
+		s.incrementSpamCount(session, validator)
+	}
+
+	s.unconfirmed[unconfirmedDisputesKey(session, candidate)] = validators
+
+	return true
+}
+
+func (s *spamSlots) Clear(session parachain.SessionIndex, candidate common.Hash) {
+	if validators, ok := s.getValidators(session, candidate); ok {
+		validatorSet := validators.Values()
+		s.clearValidators(session, candidate)
+
+		for _, validator := range validatorSet {
+			spamCount, ok := s.getSpamCount(session, validator.(parachain.ValidatorIndex))
```

Comment:
`getSpamCount` could return just the value 

```go
for _, validator := range validatorSet {
    spamCount := s.getSpamCount(session, validator.(parachain.ValidatorIndex))
    switch {
    case spamCount < 1:
    case spamCount == 1:
        s.clearSlots(session, validator.(parachain.ValidatorIndex))
    default:
        s.slotsLock.Lock()
        s.slots[slotsKey(session, validator.(parachain.ValidatorIndex))] = spamCount - 1
        s.slotsLock.Unlock()
    }
}
```

Reply:
The reason was to identify if there are no spam slots for a set of session and validator index. 

In the sample code you shared, if there is no spam slot for a set of session and validator-index, it would set the spamCount to -1 since the default return would be 0. 

We could prolly add a case to check if it is 0 and do nothing, but we won't know if that key exists or not. 

I feel keeping it as is makes the `getSpamCount()` generic ( in case we wanna know that in the future ) and keeps the code clean and readable.

Reply:
@kanishkatn in the sample code there is the case `case spamCount < 1` so if the `spamCount` is 0 that case take place and since there is nothing there then it will just jump to the next iteration, it only subtract one if the `spamCount` is greater than 1.

If you want to keep it as it is I would suggest an early continue
```go
spamCount, ok := s.getSpamCount(session, validator.(parachain.ValidatorIndex))
if !ok { continue }
...
```

Reply:
Yes, you're right and I'm blind, haha.

I'll keep it as is and do an early return. Thank you!